drivers/gpio/gpio_axp192.c

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright (c) 2023 Martin Kiepfer
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT x_powers_axp192_gpio
+
+#include <errno.h>
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/toolchain.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/mfd/axp192.h>
+
+LOG_MODULE_REGISTER(gpio_axp192, CONFIG_GPIO_LOG_LEVEL);
+
+struct gpio_axp192_config {
+	struct gpio_driver_config common;
+	struct i2c_dt_spec i2c;
+	const struct device *mfd;
+	uint32_t ngpios;
+};
+
+struct gpio_axp192_data {
+	struct gpio_driver_data common;
+	struct k_mutex mutex;
+	sys_slist_t cb_list_gpio;
+};
+
+static int gpio_axp192_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	int ret;
+	uint8_t port_val;
+	const struct gpio_axp192_config *config = dev->config;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	ret = mfd_axp192_gpio_read_port(config->mfd, &port_val);
+	if (ret == 0) {
+		*value = port_val;
+	}
+
+	return ret;
+}
+
+static int gpio_axp192_port_set_masked_raw(const struct device *dev, gpio_port_pins_t mask,
+					   gpio_port_value_t value)
+{
+	int ret;
+	const struct gpio_axp192_config *config = dev->config;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	ret = mfd_axp192_gpio_write_port(config->mfd, value, mask);
+
+	return ret;
+}
+
+static int gpio_axp192_port_set_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	return gpio_axp192_port_set_masked_raw(dev, pins, pins);
+}
+
+static int gpio_axp192_port_clear_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	return gpio_axp192_port_set_masked_raw(dev, pins, 0);
+}
+
+static int gpio_axp192_configure(const struct device *dev, gpio_pin_t pin,
+					gpio_flags_t flags)
+{
+	const struct gpio_axp192_config *config = dev->config;
+	int ret;
+	enum axp192_gpio_func func;
+
+	if (pin >= config->ngpios) {
+		LOG_ERR("Invalid gpio pin (%d)", pin);
+		return -EINVAL;
+	}
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	/* Configure pin */
+	LOG_DBG("Pin: %d / flags=0x%x", pin, flags);
+	if ((flags & GPIO_OUTPUT) != 0) {
+
+		/* Initialize output function */
+		func = AXP192_GPIO_FUNC_OUTPUT_LOW;
+		if ((flags & GPIO_OPEN_DRAIN) != 0) {
+			func = AXP192_GPIO_FUNC_OUTPUT_OD;
+		}
+		ret = mfd_axp192_gpio_func_ctrl(config->mfd, dev, pin, func);
+		if (ret != 0) {
+			return ret;
+		}
+
+		/* Set init value */
+		if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
+			ret = mfd_axp192_gpio_write_pin(config->mfd, pin, false);
+		} else if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
+			ret = mfd_axp192_gpio_write_pin(config->mfd, pin, true);
```

Comment:
Is this the only place you are using mfd_axp192_gpio_write_pin?

If so you could replace it with a call to `mfd_axp192_gpio_write_port(config->mfd, BIT(pin),...)` and remove mfd_axp192_gpio_write_pin entirely.